### ISSUE TITLE
MPI4py 4 and `get_config()`

### ISSUE THREAD

**Issue by PhilipVinc** [mpi4py/mpi4py#532]
Hello,

Over at [mpi4jax](https://github.com/mpi4jax/mpi4jax) , together with @dionhaefner, we have developed a small wrapper for MPI to use MPI4PY with jax. It works by providing some new jax primitives that work well with mpi4py communicator objects.

We need to build a small Cython module linking to the same MPI library that mpi4py was linked to.

we have been relying for a few years on `mpi4py.get_config()` to ensure that we link to exactly the same library, but in mpi4py 4.0 this was removed.

Is there a way to expose the same information, or the removal of that was deliberate, and we should fall back to using mpicc?

**Comment by dalcinl** [comment]
> Is there a way to expose the same information, or the removal of that was deliberate

The removal was deliberate. We are now distributing mpi4py wheels that can work with multiple MPI implementations, selecting the MPI library at runtime depending on what can be found automatically or via environment variables. Under this much more dynamic model, the `mpi4py.get_config()` output is no longer relevant.

Maybe the following code can help you guys to get the full path to the MPI library:

```python

import ctypes
import os
from mpi4py import MPI

class DL_Info(ctypes.Structure):
  _fields_ = [
      ("dli_fname", ctypes.c_char_p),
      ("dli_fbase", ctypes.c_void_p),
      ("dli_sname", ctypes.c_char_p),
      ("dli_saddr", ctypes.c_void_p),
  ]

libc = ctypes.CDLL(None)
dladdr = libc.dladdr
dladdr.restype = ctypes.c_int
dladdr.argtypes = [ctypes.c_void_p, ]

module = ctypes.CDLL(MPI.__file__)
symbol = module.MPI_Init
dl_info = DL_Info()
dladdr(symbol, ctypes.byref(dl_info))
libmpi = os.fsdecode(dl_info.dli_fname)

print(libmpi)
```

**Comment by dionhaefner** [comment]
Interesting. The pattern we use is something like this:

```cython
from mpi4py.libmpi cimport MPI_Comm, MPI_Comm_rank, ...
```

To build this Cython extension we need access to the `mpi.h` header that's used to build mpi4py, not the library. In the past we used `mpi4py.get_config()["mpicc"]` to find the MPI compiler that includes the correct headers. How is that solves in mpi4py now?

**Comment by dalcinl** [comment]
> How is that solved in mpi4py now?

It cannot be solved from the mpi4py side, simply because mpi4py can very well run with an MPI library that is not the same used at link time, as long as the library is ABI compatible (like MPICH and derivatives). For example, you can build mpi4py with let say MPICH 3.x, and eventually switch to Intel MPI by just running `source /opt/intel/oneapi/setvars.sh`.
In the near future, once the new MPI ABI proposal lands in MPI 4.2 or MPI 5.0, the decoupling of build time and run time will be strengthened even more, as the build may very well be done with a stubs MPI header and library, as currently under CI testing here: https://github.com/mpi4py/mpi4py-testing/actions/workflows/abi.yml

The only obvious recommendation I have is to look yourself for the `mpicc` command using `shlex.which()`, maybe also honoring a `MPICC` environment variable. The latest mpi4py release has some machinery in place to warn if mpi4py is being ran with the wrong mpiexec, so that helps to prevent users messing up they environment.

**Comment by dionhaefner** [comment]
Okay thanks. So to make sure I got that right: the headers used at build time (e.g. to generate PyPI wheels) are not part of the distribution, and neither is meta-information on which headers were used?

**Comment by dalcinl** [comment]
> the headers used at build time (e.g. to generate PyPI wheels) are not part of the distribution

Indeed. They cannot be part of the distribution, because they belong to an external dependency (MPI). Moreover, these wheels work with either MPICH or Open MPI and multiple versions of them. How is that? Brute force: there are multiple extension modules in the wheel, each built  for a each different MPI implementation and MPI standard version, and hooks on the Python import machinery select the proper extension module to match the MPI library available at runtime.

Minor clarification: The wheels at PyPI are Windows-only. Wheels for all platform are at an alternate package index at https://anaconda.org/mpi4py. I'm still hesitant to publish these wheels on the main PyPI index. 

> neither is meta-information on which headers were used?

Exactly. However, you have `MPI.Get_library_version()` to figure out some details of the MPI implementation at runtime. 
Again, my recommendation is to "trust" the build environment, and use `mpicc` as found in `$PATH` (eg. by using `os.environ.get("MPICC") or shutil.which("mpicc")`), otherwise let the Python build system use whatever compiler , assuming the MPI header and libraries are found automatically by `$CC`. That's mostly the way mpi4py builds itself.

**Comment by dionhaefner** [comment]
Sounds good, thank you very much!
